lib/route/cls/flower.c

```diff
@@ -598,6 +649,126 @@ int rtnl_flower_get_ip_dscp(struct rtnl_cls *cls, uint8_t *dscp, uint8_t *mask)
 	return 0;
 }
 
+/**
+ * Set IPv4 source address for flower classifier
+ * @arg cls		Flower classifier.
+ * @arg addr		IPv4 source address
+ * @arg mask		mask for IPv4 source address
+ * @return 0 on success or a negative error code.
+ */
+int rtnl_flower_set_ipv4_src(struct rtnl_cls *cls, in_addr_t addr, in_addr_t mask)
+{
+	struct rtnl_flower *f;
+
+	if (!(f = rtnl_tc_data(TC_CAST(cls))))
+		return -NLE_NOMEM;
+
+	if (addr) {
+		f->cf_ipv4_src = htonl(addr);
+		f->cf_mask |= FLOWER_ATTR_IPV4_SRC;
+
+		if (mask) {
+			f->cf_ipv4_src_mask = htonl(mask);
+			f->cf_mask |= FLOWER_ATTR_IPV4_SRC_MASK;
+		}
+
+		return 0;
+	}
+
+	return -NLE_FAILURE;
+}
+
+/**
+ * Get IPv4 source address for flower classifier
+ * @arg cls		Flower classifier.
+ * @arg addr		IPv4 source address
+ * @arg mask		mask for IPv4 source address
+ * @return 0 on success or a negative error code.
+ */
+int rtnl_flower_get_ipv4_src(struct rtnl_cls *cls, in_addr_t *out_addr,
+			     in_addr_t *out_mask)
+{
+	struct rtnl_flower *f;
+
+	if (!(f = rtnl_tc_data_peek(TC_CAST(cls))))
+		return -NLE_INVAL;
+
+	if (!(f->cf_mask & FLOWER_ATTR_IPV4_SRC))
+		return -NLE_MISSING_ATTR;
+
+	if (out_addr)
+		*out_addr = ntohl(f->cf_ipv4_src);
+
+	if (out_mask) {
+		if (f->cf_mask & FLOWER_ATTR_IPV4_SRC_MASK)
+			*out_mask = ntohl(f->cf_ipv4_src_mask);
+		else
+			*out_mask = 0xffffffff;
+	}
+
+	return 0;
+}
+
+/**
+ * Set IPv4 destination address for flower classifier
+ * @arg cls		Flower classifier.
+ * @arg addr		IPv4 destination address
+ * @arg mask		mask for IPv4 destination address
+ * @return 0 on success or a negative error code.
+ */
+int rtnl_flower_set_ipv4_dst(struct rtnl_cls *cls, in_addr_t addr, in_addr_t mask)
+{
+	struct rtnl_flower *f;
+
+	if (!(f = rtnl_tc_data(TC_CAST(cls))))
+		return -NLE_NOMEM;
+
+	if (addr) {
+		f->cf_ipv4_dst = htonl(addr);
+		f->cf_mask |= FLOWER_ATTR_IPV4_DST;
+
+		if (mask) {
+			f->cf_ipv4_dst_mask = htonl(mask);
+			f->cf_mask |= FLOWER_ATTR_IPV4_DST_MASK;
+		}
+
+		return 0;
+	}
```

Comment:
this does not seem right. The current code does:

- fail if `addr==0` and do nothing. Why?
- only touch mask if `mask!=0`. Why?

I think the setters should not go overboard with validity checking. Especially if it's done inconsistently.

- Is `mask==0` something disallowed by kernel? If so, we could use that to unset `FLOWER_ATTR_IPV4_DST_MASK`.
- Is `addr==0` something disallowed by kernel? If so, we could use that to unset `FLOWER_ATTR_IPV4_DST`.

Otherwise, the setter cannot express to unset `cf_mask`, and still support setting any value, uncluding `0.0.0.0`. Not being able to unset it is not nice (if having it unset is a possible configuration that can be achieved otherwise).